### PR TITLE
Generalizing WFRC's Tool

### PR DESCRIPTION
**Pull request by jacobdadams**
This big, oversized PR reworks WFRC's tool that Josh created for Davis County (and subsequently modified for other WFRC counties). It converts it from jupyter notebooks to a standalone program, refactors it into methods and modules for clarity, and creates a basic framework for writing semi-custom scripts for other counties. 

The aim is to create an interface that other county's specific data can be transformed to so that the rest of the script for each county is relatively simple code reuse.

`docs/refactor_notes.md` has some info on the process.

There are a few remaining tasks that will be addressed in the near future:

1. Trimming the history to remove the GDBs so that the repo shrinks significantly.
2. Implementing the UNITID field.
3. Some higher-level process questions for Josh and company.

I forked this so that it's clear this is a derivative of Josh's work and points back to their efforts. If you have suggestions for cleaner ways to accomplish that, I'm all ears.

### CODE REVIEW

**Comment by jacobdadams** [comment]
Whoops, belay that. Need to get CI set up for tests.

**Comment by jacobdadams** [comment]
Ok, CI is wired up.

File: README.md
Comment (steveoh) [inline README.md]: **question**: how can the inventory use a custom created mobile home community layer?

File: README.md
Comment (steveoh) [inline README.md]:
```suggestion
- **UNIT_COUNT**: Number of units in the feature. Set manually for single family homes and duplexes and for all others (owned unit groupings, mobile homes, multi-family) it uses the address point count
```

File: .github/workflows/ci.yml
Comment (stdavis) [inline .github/workflows/ci.yml]:
```suggestion
      uses: actions/checkout@v3
```
**nitpick**

File: .vscode/settings.json
Comment (stdavis) [inline .vscode/settings.json]:
```suggestion
  "python.testing.pytestEnabled": true,
```
**nitpick**

File: README.md
Comment (stdavis) [inline README.md]:
```suggestion
- **UNIT_COUNT**: Number of units in the feature. Set manually for single family homes and duplexes and for all others (owned unit groupings, mobile homes, multi-family) it uses the address point count
```

File: src/housing_unit_inventory/county_scripts.py
Comment (stdavis) [inline src/housing_unit_inventory/county_scripts.py:19]:
```suggestion
```

File: src/housing_unit_inventory/helpers.py
Comment (stdavis) [inline src/housing_unit_inventory/helpers.py:69]: Should this be removed?

**Review by stdavis — CHANGES_REQUESTED** [review]
There are several `FIXME` comments. Did you want to resolve those before a merge or punt them until later?

File: src/housing_unit_inventory/calculate.py
Comment (stdavis) [inline src/housing_unit_inventory/calculate.py:21]: **question**: Is there a reason why you are using bitwise operators rather than the `or`/`and` boolean operators? I always thought that bitwise was for bits or numbers but I could be wrong. I think that the outcome would be similar, if not exactly the same, for this code. I'm just wondering if I'm missing something.

File: setup.py
Comment (steveoh) [inline setup.py:15]: **question**: this should be set to something or is intentionally blank?

File: src/housing_unit_inventory/calculate.py
Comment (jacobdadams) [reply src/housing_unit_inventory/calculate.py:21]: Within a pandas `.loc` statement, you have to use `| &` instead of or/and. The first line there was me pulling a complex statement out into its own variable... not sure if it makes a difference now that it's outside of a `.loc`...

File: src/housing_unit_inventory/county_scripts.py
Comment (jacobdadams) [reply src/housing_unit_inventory/county_scripts.py:19]: Leaving this in for now (it's a drastically smaller subset of parcels for quick test runs)

File: src/housing_unit_inventory/calculate.py
Comment (steveoh) [inline src/housing_unit_inventory/calculate.py:16]: **suggestion**: create an issue for this?

File: src/housing_unit_inventory/calculate.py
Comment (steveoh) [inline src/housing_unit_inventory/calculate.py:65]: **praise**: this is pretty cool thanks for the comment otherwise i'd have no idea.

File: src/housing_unit_inventory/calculate.py
Comment (steveoh) [inline src/housing_unit_inventory/calculate.py:85]: **suggestion**: create an issue

File: src/housing_unit_inventory/county_scripts.py
Comment (steveoh) [inline src/housing_unit_inventory/county_scripts.py:34]: **suggestion**: create a cli for these and use some relative path conventions to simplify

File: src/housing_unit_inventory/davis2020.py
Comment (steveoh) [inline src/housing_unit_inventory/davis2020.py:22]: **suggestion**: create issue

File: src/housing_unit_inventory/davis2020.py
Comment (steveoh) [inline src/housing_unit_inventory/davis2020.py:522]: **question**: is there a reason you pad your queries with spaces?

File: src/housing_unit_inventory/davis2020.py
Comment (steveoh) [inline src/housing_unit_inventory/davis2020.py:677]: **chore**: refactor to cli

File: src/housing_unit_inventory/davis2020.py
Comment (steveoh) [inline src/housing_unit_inventory/davis2020.py:827]: **thought**: it seems odd to be doing bitwise and here with booleans. I thought I also saw a function that calculates the decade. Shouldn't that be used here?

File: src/housing_unit_inventory/davis2020.py
Comment (steveoh) [inline src/housing_unit_inventory/davis2020.py:851]: **thought**: it is confusing to use `~` instead of negating the condition with`!=`. Is there a specific reason worth sharing why this way was chosen?

File: src/housing_unit_inventory/dissolve.py
Comment (steveoh) [inline src/housing_unit_inventory/dissolve.py:21]: **question**: what is special about 2 here? Why wouldn't the method below work regardless of the size? If 2 is special would you put it in a named variable?

File: src/housing_unit_inventory/helpers.py
Comment (steveoh) [inline src/housing_unit_inventory/helpers.py:267]: **nitpick**: if this is a boolean answer of yes or no, a boolean would make more sense.

**Review by steveoh — APPROVED** [review]
This is a lot to review as you mentioned. I noticed a few things worth sorting out. 

1. make this obvious and easy to run. I'm not sure how i'd try to run it myself right now. There are a lot of geodatabases and files that need to be in specific places and I'm not sure how or what those are supposed to be. Maybe some hand holding in a cli like [inquirer](https://pypi.org/project/inquirer/) (_I'm not sure if that's the best choice these days but you can see the point._) 
2. Organize the print statements, logging statements, and warning statements. I'd try to be consistent, and replace the prints with logging. I'm not familiar with warning but it seems fine and similar to logging.warn so I'm not sure why it's necessary.
3. Turn all the fixme's and todo's etc into issues. There's a lot more work to be done but it would be silly in my opinion to hold up the pr until that was done. Let's get the history into the main or dev branch. I might recommend editing the base of this pr to be `dev` since it may not be ready for production yet? I'll leave that up to you. 
4. Let's get dependabot.yml in here as well that will keep the actions and python current.

File: src/housing_unit_inventory/davis2020.py
Comment (jacobdadams) [reply src/housing_unit_inventory/davis2020.py:22]: My bad for not explaining- `davis2020.py` was a first shot at translating the notebook to a script and then switching from arcpy to arcgis api for python. This is left for reference for now but can probably be deleted later.

File: src/housing_unit_inventory/davis2020.py
Comment (jacobdadams) [reply src/housing_unit_inventory/davis2020.py:522]: ibid. Also, this could have been copy/pasted from Josh's notebooks, don't remember.

File: src/housing_unit_inventory/davis2020.py
Comment (jacobdadams) [reply src/housing_unit_inventory/davis2020.py:677]: ibid.

File: src/housing_unit_inventory/davis2020.py
Comment (jacobdadams) [reply src/housing_unit_inventory/davis2020.py:827]: ibid. Also, pandas uses `& |` within `.loc` statements instead of `and or`. So here, they're not bitwise.

File: src/housing_unit_inventory/davis2020.py
Comment (jacobdadams) [reply src/housing_unit_inventory/davis2020.py:851]: ibid. Also, I don't remember why. This could have been the result of fussing around to get the right pandas expression, or it could have been copied from the notebooks. The updated davis county method in `county_scripts.py` uses `calculate.remove_zero_unit_house_counts` instead, which is cleaner and more robust.

File: src/housing_unit_inventory/dissolve.py
Comment (jacobdadams) [reply src/housing_unit_inventory/dissolve.py:21]: So, this recursively calls itself until there are only two shapes left, and then works its way back up the stack dissolving the base case two shapes and then that resulting shape with the next shape in the series, and so on back up to the top. I can't think of a good, concise variable name to encapsulate that- would a comment be better here?

File: src/housing_unit_inventory/helpers.py
Comment (jacobdadams) [reply src/housing_unit_inventory/helpers.py:267]: This value gets written to the final feature class as a note to the end user whether the parcel is an OUG or just a normal parcel. The field is currently a string field (based on WFRC's first data submission), so I'd have to convert this to a string at some point before writing it out and `Yes/No` felt less programmy than `True/False`.

File: src/housing_unit_inventory/helpers.py
Comment (jacobdadams) [reply src/housing_unit_inventory/helpers.py:69]: added issue #3

File: setup.py
Comment (jacobdadams) [reply setup.py:15]: Thanks for the reminder. Need to verify with WFRC since I'm (heavily) modifying their code/process.

File: README.md
Comment (jacobdadams) [reply README.md]: Reworded.

File: src/housing_unit_inventory/dissolve.py
Comment (steveoh) [reply src/housing_unit_inventory/dissolve.py:21]: Oh I should have picked up on that based on the name. I didn't see the tail recursion call and probably skimmed it wherever it is. A comment would be good.

**Comment by jacobdadams** [comment]
Thanks for the reviews; I've made changes as noted. @steveoh, to your points:

1. Yeah, it's still a bit of a mess. It's kind of a library and client rolled into one. Each county is unique enough that they'll need their own custom script/function, which I'm currently keeping in `county_scripts.py`. The other modules (`calculate.py`, `dissolve.py`, `evaluations.py`, and `helpers.py`) hold different, reusable parts of the process. `davis2020.py` was a first translation from notebook to script and arcpy to arcgis api for python. In the future, it may make sense to have individual modules or packages for each county with their own settings files and cli entry points, or some other way to call them, but for now it's still pretty custom/hard coded. There's also a lot of input data that live in the repo that needs to be re-homed.
2. All the new/update stuff uses logging; the print statements should just be in the old prototype code. As for the warnings, the [python docs](https://docs.python.org/3/howto/logging.html) recommend using `warnings.warn` for situations the client code can do something about and `logging.warn` for things they can't. The currently-used `warnings.warn` are for situations where the input data could/should be modified.
3. That makes sense, and I've switched it around.
4. I couldn't find a good python dependabot example in our org, so I grabbed one online. Take a look and let me know if looks good to you. We should probably also put this in the `agrc/python` template as well.

**Comment by steveoh** [comment]
> 1. It's kind of a library and client rolled into one.

Yeah, that makes sense. It could be interesting to have another person in the office try to run it and see what happens. Any tools to create data or to populate pro projects yada yada for the process to run would greatly help. It's tough on something like this that is so specific and complicated so anything to help future us is a good idea. 

> 2. ... use warnings.warn for situations the client code can do something about and logging.warn for things they can't. 

This is great advice. Thanks for sharing.

> 4\. couldn't find a good python dependabot example in our org, so I grabbed one online. Take a look and let me know if looks good to you. We should probably also put this in the `agrc/python` template as well.

There is a start in [agrc/python](https://github.com/agrc/python/blob/main/.github/dependabot.yml). It could use the actions. A pr is welcome.

Otherwise what you have looks good. It seems the same as the [open-sgid](https://github.com/agrc/open-sgid/blob/main/.github/dependabot.yml).

Sounds like a lot of our other python projects could use the file. Feel free to create issues for those or pr's whatever or we'll sort it out later.

**Comment by jacobdadams** [comment]
> Yeah, that makes sense. It could be interesting to have another person in the office try to run it and see what happens. Any tools to create data or to populate pro projects yada yada for the process to run would greatly help. It's tough on something like this that is so specific and complicated so anything to help future us is a good idea.

Definitely. At the very least, documentation will be forthcoming. Figuring out the inputs has been one of the tasks :)
 
> There is a start in [agrc/python](https://github.com/agrc/python/blob/main/.github/dependabot.yml). It could use the actions. A pr is welcome.
> Sounds like a lot of our other python projects could use the file. Feel free to create issues for those or pr's whatever or we'll sort it out later.

I'm either blind or forgot to do a pull on my local copy.

**Review by stdavis — APPROVED** [review]

**Pull request merged**
